pkg/agent/plugin/nodeattestor/k8s/k8s.go

```diff
@@ -0,0 +1,245 @@
+package k8s
+
+import (
+	"context"
+	"crypto"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+	"path"
+	"sync"
+
+	"github.com/hashicorp/hcl"
+	"github.com/spiffe/spire/pkg/common/plugin/k8s"
+	"github.com/spiffe/spire/pkg/common/plugin/x509pop"
+	"github.com/spiffe/spire/proto/agent/nodeattestor"
+	"github.com/spiffe/spire/proto/common"
+	"github.com/spiffe/spire/proto/common/plugin"
+
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/util/cert"
+	"k8s.io/client-go/util/certificate/csr"
+)
+
+const (
+	pluginName = "k8s"
+)
+
+type configData struct {
+	spiffeID        string
+	privateKey      crypto.PrivateKey
+	attestationData *common.AttestationData
+}
+
+type K8sConfig struct {
+	TrustDomain     string `hcl:"trust_domain"`
+	PrivateKeyPath  string `hcl:"k8s_private_key_path"`
+	CertificatePath string `hcl:"k8s_certificate_path"`
+	K8sCACertPath   string `hcl:"k8s_ca_certificate_path"`
+	KubeconfigPath  string `hcl:"kubeconfig_path"`
+}
+
+type K8sPlugin struct {
+	m          sync.Mutex
+	c          *K8sConfig
+	kubeClient kubernetes.Interface
+}
+
+var _ nodeattestor.Plugin = (*K8sPlugin)(nil)
+
+func (p *K8sPlugin) getConfig() *K8sConfig {
+	p.m.Lock()
+	defer p.m.Unlock()
+	return p.c
+}
+
+func (p *K8sPlugin) setConfig(c *K8sConfig) {
+	p.m.Lock()
+	defer p.m.Unlock()
+	p.c = c
+}
+
+func getAgentName() string {
+	name, err := os.Hostname()
+	if err != nil {
+		name = "unknown"
+	}
+	return name
+}
+
+func getKubeClient(kubeConfigFilePath, clientCertFilePath, clientKeyFilePath, caCertFilePath string) (kubernetes.Interface, error) {
+	if kubeConfigFilePath == "" {
+		// Try KUBECONFIG env variable
+		kubeConfigFilePath = os.Getenv("KUBECONFIG")
+		if kubeConfigFilePath == "" {
+			// Still no luck, try default (home)
+			home := os.Getenv("HOME")
+			if home != "" {
+				kubeConfigFilePath = path.Join(home, ".kube", "config")
+			}
+		}
+	}
+
+	if kubeConfigFilePath == "" {
+		return nil, fmt.Errorf("Unable to locate kubeconfig")
+	}
+
+	config, err := clientcmd.BuildConfigFromFlags("", kubeConfigFilePath)
+	if err != nil {
+		return nil, fmt.Errorf("Error accessing kubeconfig %s: %v", kubeConfigFilePath, err)
+	}
+
+	config.TLSClientConfig.CertFile = clientCertFilePath
+	config.TLSClientConfig.KeyFile = clientKeyFilePath
+	config.TLSClientConfig.CAFile = caCertFilePath
+
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, fmt.Errorf("Error creating clientset: %v", err)
+	}
+	return clientset, nil
+}
+
+func fetchK8sCert(kubeClient kubernetes.Interface) (*tls.Certificate, error) {
+	key, err := cert.MakeEllipticPrivateKeyPEM()
+	if err != nil {
+		return nil, fmt.Errorf("Error creating private key: %v", err)
+	}
+
+	certsIntf := kubeClient.CertificatesV1beta1().CertificateSigningRequests()
```

Comment:
I noticed the `beta1` suffix here - how stable is this API? Do we expect it to be the same a year or more from now?

Reply:
The feature that depends on this API is going GA in the next release, so in the long term the mechanics of node attestation should be stable.
The API itself will not graduate yet, though, so minor changes are possible.
This is the issue for tracking and discussion:

https://github.com/kubernetes/features/issues/43#issuecomment-411968090